pkg/controllers/provisioner/v1alpha1/allocation/greedyallocator.go

```diff
@@ -45,25 +46,69 @@ func (a *GreedyAllocator) Allocate(pods []*v1.Pod) error {
 
 type SchedulingGroup struct {
 	Pods        []*v1.Pod
-	Constraints cloudprovider.CapacityConstraints
+	Constraints *cloudprovider.CapacityConstraints
 }
 
-func (a *GreedyAllocator) getSchedulingGroups(pods []*v1.Pod) []SchedulingGroup {
-	groups := []SchedulingGroup{}
-
+func (a *GreedyAllocator) getSchedulingGroups(pods []*v1.Pod) []*SchedulingGroup {
+	groups := []*SchedulingGroup{}
 	for _, pod := range pods {
+		added := false
 		for _, group := range groups {
-			if a.matchesGroup(pod, group) {
-				group.Pods = append(group.Pods, pod)
+			if a.matchesGroup(group, pod) {
+				a.addToGroup(group, pod)
+				added = true
 				break
 			}
 		}
+		if added {
+			continue
+		}
+		newGroup := newSchedulingGroup()
+		a.addToGroup(newGroup, pod)
+		groups = append(groups, newGroup)
 	}
-
 	return groups
 }
 
+func newSchedulingGroup() *SchedulingGroup {
+	return &SchedulingGroup{
+		Constraints: &cloudprovider.CapacityConstraints{
+			Resources: v1.ResourceList{
+				v1.ResourceCPU:    resource.MustParse("0"),
+				v1.ResourceMemory: resource.MustParse("0"),
+			},
+		},
+	}
+}
+
+func (a *GreedyAllocator) addToGroup(group *SchedulingGroup, pod *v1.Pod) {
```

Comment:
Should this function just be a constructor?

func SchedulingGroupFor(pod *v1.Pod) SchedulingGroup {}

This will allow you to initialize all of the group's parameters without mutating them.

Reply:
Makes sense, I will make that change.